### PR TITLE
Added iOS support for including images in a bundle other than the mainBundle

### PR DESCRIPTION
**Pull request by mpretty-homepass**
This allows a developer to specify a 'bundle' when defining the source for an Image as shown below. If a bundle was specified the code will loop through all bundles looking for one with that name, if found it will use that bundle. In all other cases (if the bundle wasn't found or no bundle was specified) it will default to using the mainBundle.

```
<Image source={{ uri: 'image', bundle: 'BundleName' }} />
```

Closes https://github.com/facebook/react-native/issues/2282 (or https://productpains.com/post/react-native/image-loading-from-named-bundles-uiimage-imagenamedinbundlecompatiblewithtraitcollection since the issue was moved to ProductPains)

### CODE REVIEW

**Comment by mpretty-homepass** [comment]
Looks like I broke non-bundled images. Will update with a fix in a minute.

**Comment by nicklockwood** [comment]
Thanks, this looks like a good idea. But can you include an example of loading an image from a secondary bundle in the UIExplorer image example?

**Comment by mpretty-homepass** [comment]
@nicklockwood I've added the example to UIExplorer image example, fixed the error from the last build and tweaked the logic to allow for people who directly include Resource bundles in their project (previously it was only working for bundles installed via CocoaPods).

**Comment by mpretty-homepass** [comment]
Damn, looks like I need to update the tests? Will update in the morning.

**Comment by nicklockwood** [comment]
Might just need to re-record the snapshot image. I'll see if I can fix it internally.

@facebook-github-bot shipit

**Comment by mpretty-homepass** [comment]
@nicklockwood I've fixed the tests which broke due to my RCTImageLoader changes but there are some others which don't seem to be related.

**Comment by mpretty-homepass** [comment]
Sorry, that will hopefully be the last commit - looks like it wasn't actually doing what I needed it to do (load assets from a bundle included in a CocoaPod framework). I've tweaked the code to also search through all bundles included in each framework returned by [NSBundle allFrameworks].

**Comment by mpretty-homepass** [comment]
@nicklockwood Were you able to figure out what was going on with the tests?

I also just found an edge cases where frameworks might not actually have a 'infoDictionary' so have added a code to handle it.

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit

**Comment by mpretty-homepass** [comment]
@nicklockwood Do you know what's going on with the tests? (I'm assuming the below errors aren't related to my changes)

```
<unknown>:0: failed: caught "NSInternalInconsistencyException", "*** Assertion failure in -[RCTTestRunner runTest:module:initialProps:configurationBlock:expectErrorBlock:], /Users/travis/build/facebook/react-native/Libraries/RCTTest/RCTTestRunner.m:154: RedBox error: Could not connect to development server.
```
```
<unknown>:0: failed: caught "NSInternalInconsistencyException", "*** Assertion failure in -[RCTTestRunner runTest:module:initialProps:configurationBlock:expectErrorBlock:], /Users/travis/build/facebook/react-native/Libraries/RCTTest/RCTTestRunner.m:154: RedBox error: Unhandled JS Exception: _fail called by testOrdering-t50, setTimeout 20 should happen before setTimeout 50"
```

**Comment by mpretty-homepass** [comment]
@nicklockwood I still seem to be getting the timeout issues in the tests, any idea what's going on?

**Comment by facebook-github-bot-1** [comment]
@a2 would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by geirman** [comment]
@mpretty-homepass @nicklockwood looks like we lost momentum on this, but we're super close. What's needed to kick-start this again? I assume a rebase at the very least.

Awesome work on a useful feature!

**Comment by spikebrehm** [comment]
We're also in need of this functionality at Airbnb. How can we help?

**Comment by facebook-github-bot-9** [comment]
@mpretty-homepass updated the pull request.

**Comment by mpretty-homepass** [comment]
Hey @geirman, @spikebrehm, @nicklockwood, sorry for the massive delay - things have been pretty crazy at work over the past few months. I've updated the PR to the latest master and it all seems to be working - let me know if you run in to any issues.

**Comment by facebook-github-bot-9** [comment]
@mpretty-homepass updated the pull request.

**Comment by spikebrehm** [comment]
> sorry for the massive delay - things have been pretty crazy at work over the past few months

@mpretty-homepass no worries! I've been there. Let's try to get some FB eyes on this PR.

**Comment by spikebrehm** [comment]
@mpretty-homepass it does look that there's a merge conflict now.

File: Libraries/Image/RCTLocalAssetImageLoader.m [hunk 1]
```diff
@@ -50,7 +51,49 @@ - (RCTImageLoaderCancellationBlock)loadImageForURL:(NSURL *)imageURL
...
-    UIImage *image = [UIImage imageNamed:imageName];
+    NSBundle *targetBundle = [NSBundle mainBundle];
+    
+    if (bundlePath) {
+      NSString *targetBundleName = RCTBundlePathForURL([NSURL URLWithString:bundlePath]);
+      NSString *targetBundleNameWithExtension = [NSString stringWithFormat:@"%@%@", targetBundleName, @".bundle"];
+      NSMutableArray *bundles = [[NSBundle allBundles] mutableCopy];
+      
+      // Get any bundle files included in the mainBundle and add them to the array
+      for (NSString *path in [[NSBundle mainBundle] pathsForResourcesOfType:@"bundle" inDirectory:nil]) {
```
Comment (spikebrehm) [inline Libraries/Image/RCTLocalAssetImageLoader.m]: I don't know much about iOS development, but it seems like we're iterating through every bundle and every framework every time `loadImageForURL` is called.  Would it be possible to cache this work and only do it the first time?  Or is it possible that the bundles and frameworks can change at runtime?

**Comment by lelandrichardson** [comment]
ping @nicklockwood would be great for us if we could get this in the next release! :)

**Comment by facebook-github-bot-9** [comment]
@mpretty-homepass updated the pull request.

**Comment by mpretty-homepass** [comment]
@spikebram I can't see any conflicts - have updated to the latest master anyway

File: Libraries/Image/RCTLocalAssetImageLoader.m (hunk 1 above)
Comment (mpretty-homepass) [inline Libraries/Image/RCTLocalAssetImageLoader.m]: No, they can't change at runtime so this should be cached; I'll see what I can plug in quickly.

**Comment by facebook-github-bot-9** [comment]
@mpretty-homepass updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@mpretty-homepass updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@mpretty-homepass updated the pull request - [view changes](https://github.com/facebook/react-native/pull/5868/files/e10bfa52373d607e8e5762f8eb41ac12117b72e6..820557a9602c403e1a771c7eba157d207f71624e)

File: Libraries/Image/RCTLocalAssetImageLoader.m (hunk 1 above)
Comment (mpretty-homepass) [inline Libraries/Image/RCTLocalAssetImageLoader.m]: Added a map to keep track of bundles which have been loaded; while they can't change at runtime they can change between loads.

It also doesn't really need to be more complicated than this because (as far as I can tell) it doesn't load the entire bundle in to memory when storing it.

File: Examples/UIExplorer/UIExplorerUnitTests/RCTImageLoaderHelpers.m [hunk 2]
```diff
@@ -47,9 +47,9 @@ - (BOOL)canLoadImageURL:(NSURL *)requestURL
   return _canLoadImageURLHandler(requestURL);
 }
 
-- (RCTImageLoaderCancellationBlock)loadImageForURL:(NSURL *)imageURL size:(CGSize)size scale:(CGFloat)scale resizeMode:(RCTResizeMode)resizeMode progressHandler:(RCTImageLoaderProgressBlock)progressHandler completionHandler:(RCTImageLoaderCompletionBlock)completionHandler
+- (RCTImageLoaderCancellationBlock)loadImageForURL:(NSURL *)imageURL bundlePath:(NSString *)bundlePath size:(CGSize)size scale:(CGFloat)scale resizeMode:(RCTResizeMode)resizeMode progressHandler:(RCTImageLoaderProgressBlock)progressHandler completionHandler:(RCTImageLoaderCompletionBlock)completionHandler
```
Comment (javache) [inline Examples/UIExplorer/UIExplorerUnitTests/RCTImageLoaderHelpers.m]: I feel bundlePath should be resolved at the point where we create the NSURL. That NSURL should be relative to your custom bundle.

File: Libraries/Image/RCTShadowVirtualImage.m [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+#import "RCTShadowVirtualImage.h"
+#import "RCTImageLoader.h"
+#import "RCTImageUtils.h"
+#import "RCTBridge.h"
+#import "RCTConvert.h"
+#import "RCTUIManager.h"
+#import "RCTUtils.h"
+
+@implementation RCTShadowVirtualImage
```
Comment (javache) [inline Libraries/Image/RCTShadowVirtualImage.m]: Why is this change required? Can you make it a separate PR?

File: Libraries/Image/RCTShadowVirtualImage.m [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    RCTZeroIfNaN(self.width),
+    RCTZeroIfNaN(self.height),
+  };
+
+  if (!_image) {
+    _image = RCTGetPlaceholderImage(imageSize, nil);
+  }
+
+  __weak RCTShadowVirtualImage *weakSelf = self;
+  _cancellationBlock = [_bridge.imageLoader loadImageWithTag:_source.imageURL.absoluteString
```
Comment (javache) [inline Libraries/Image/RCTShadowVirtualImage.m]: The current behaviour is for images to only load when they're actually on screen. This always load images.

File: Libraries/Image/RCTShadowVirtualImage.m (hunk 3 above)
Comment (mpretty-homepass) [inline Libraries/Image/RCTShadowVirtualImage.m]: Not sure why it looks like I added the entire file - If you looks at the file it was added in October last year (91e6c98)

File: Libraries/Image/RCTShadowVirtualImage.m (hunk 4 above)
Comment (mpretty-homepass) [inline Libraries/Image/RCTShadowVirtualImage.m]: The only change I made to this file was adding the 'bundlePath:' param to the 'loadImageWithTag:' method.

File: Examples/UIExplorer/UIExplorerUnitTests/RCTImageLoaderHelpers.m (hunk 2 above)
Comment (mpretty-homepass) [inline Examples/UIExplorer/UIExplorerUnitTests/RCTImageLoaderHelpers.m]: While I could move the bundle finding logic up to the point where the NSURL is created, we would then need to break apart the URL _[1]_ at this point in order to load it from the relevant bundle.

This would also involve a much larger functional change as now all local images would be loaded differently whereas with the current code they are only loaded differently when specifying a custom bundle path _[2]_ _[3]_.

It also adds another point of failure; the ```RCTBundlePathForURL``` is actually fairly complex and it assumes the asset is in the mainBundle; so in order to safely (assuming that method uses the correct approach) break apart the URL we would likely need to loop through all of the bundles anyway in order to recreate the above method with the ability to specify a bundle.

But seriously, if you do know of a way to load an asset from a bundle with a given URL or can think of a nicer way to do this (knowing the above), then let me know because I'd be happy to reduce the amount of code needed for this PR.

(And sorry for the massive reply)

----------------------------------------------------------------
_1._ While we might be able to load an image from another bundle using ```[UIImage initWithContentsOfFile:]``` it unfortunately doesn't seem to work for assets which have been included in an '.xcassets' file (those are built in to a .car file and cannot be accessed directly as far as I can tell). Which means we have to use the ```[UIImage imageNamed:inBundle:compatibleWithTraitCollection:]``` method.

_2._ ```[UIImage imageName:]``` just defaults to the mainBundle.

_3._ If we ignored the .xcassets case there is still an issue where the caching behaviour between [UIImage imageNamed:inBundle:compatibleWithTraitCollection:] and [UIImage imageWithContentsOfFile:] actually differ. ```imageWithContentsOfFile:``` doesn't load the image in to the system image cache which could affect performance when loading lots of reusable images rapidly [https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIImage_Class/index.html#//apple_ref/doc/uid/TP40006890-CH3-SW10](url)

**Comment by spikebrehm** [comment]
@mpretty-homepass you've [raised a good question](https://github.com/facebook/react-native/pull/5868/files/820557a9602c403e1a771c7eba157d207f71624e#r75035046)!  How can we move forward making a decision? Do we need to find someone on the RN core team to chime in?

File: Libraries/Image/RCTShadowVirtualImage.m (hunk 4 above)
Comment (javache) [inline Libraries/Image/RCTShadowVirtualImage.m]: You'll need to rebase, this file no longer exists.

**Comment by facebook-github-bot-1** [comment]
@mpretty-homepass updated the pull request - [view changes](https://github.com/facebook/react-native/pull/5868/files/820557a9602c403e1a771c7eba157d207f71624e..4d65b31f0449e1431768f8b6b89c15d49e18a6f8)

**Comment by mpretty-homepass** [comment]
@javache Updated

@spikebrehm Do you mean https://github.com/facebook/react-native/pull/5868#discussion_r74969632? The PR as it is should be good to go unless we want to look at implementing this change.

Having a single URL that evaluates to a specific bundle does seem like a good idea (as it would reduce the scope of the change); but it does mean we would need to parse the URL and invent a convention. And it would affect **all** local image loading (as we would need to parse all URLs to determine their location rather than just ones which provide a bundlePath).

I can look at reworking the logic around that idea but to me explicitly defining a 'bundlePath' seems like simpler logic than generating then parsing a URL. If we do want this change then we would probably need some input from the core RN team about how best to approach this and wether **changing how all local image locations are resolved** is a better option than only having custom load logic when the bundlePath is supplied.

**Comment by spikebrehm** [comment]
@mpretty-homepass oh ok, great!  So this is good to go, then.  

I'll ask @lelandrichardson to merge this in.

Are there any tests we can write to prevent regressions in this behavior?

And finally: why hasn't Travis reported its status yet?

**Comment by spikebrehm** [comment]
@mpretty-homepass can you rebase and push? Maybe that will kick off a new Travis build.

**Comment by spikebrehm** [comment]
PTAL @mmmulani

File: Libraries/Image/ImageSourcePropType.js [hunk 5]
```diff
@@ -22,6 +22,12 @@ const ImageURISourcePropType = PropTypes.shape({
    */
   uri: PropTypes.string,
   /**
+   * `bundle` is the iOS asset bundle which the image is included in. This
+   * will default to [NSBundle mainBundle] if not set and does not apply on
+   * Android.
```
Comment (mmmulani) [inline Libraries/Image/ImageSourcePropType.js]: add `@platform ios` here

**Comment by javache** [comment]
I'm going to put up a simple version of this that just does the URL translation and will use the file image loader. We should then change RCTBundlePathForURL to support this behaviour too, so RCTLocalAssetImageLoader can load these files too and we get OS caching.

**Comment by facebook-github-bot-6** [comment]
@mpretty-homepass updated the pull request - [view changes](https://github.com/facebook/react-native/pull/5868/files/4d65b31f0449e1431768f8b6b89c15d49e18a6f8..d5b07ae0a305932c4d701487c9249aeacb780f8c)

**Comment by mpretty-homepass** [comment]
Woops...

**Comment by facebook-github-bot-9** [comment]
@mpretty-homepass updated the pull request - [view changes](https://github.com/facebook/react-native/pull/5868/files/d5b07ae0a305932c4d701487c9249aeacb780f8c..5401c70286265cb22dd8caf4e2b87abbb1790d57)

**Comment by mpretty-homepass** [comment]
Ok, that's better. @spikebrehm It should be good now. @mmmulani Added '@platform ios'.

@javache No worries, please link the PR when it's done and let me know if you need a hand. As I said my only concern with that approach is that it would affect all image loading rather than be limited to just when specifying a bundle.

**Comment by javache** [comment]
Is there a use case here to support all bundles (e.g. loading from system framework bundles), or is just loading from the app's own bundles sufficient (with the additional feature of looking up bundles by CFBundleName)

**Comment by spikebrehm** [comment]
> or is just loading from the app's own bundles sufficient (with the additional feature of looking up bundles by CFBundleName)

This is our use case.  @mpretty-homepass, what about you?

**Comment by javache** [comment]
Here's what's being reviewed internally right now: https://gist.github.com/javache/2890602dd44c8cf27f1f24be1dd98530. Feel free to patch in locally and test.

**Comment by mpretty-homepass** [comment]
@javache Yep, that is much more succinct than my implementation.

The main differences being:
- All bundles cached once the first file is loaded from a bundle (mine was only caching the specific bundle used)
- The bundle name is prefixed on to the imageURL, in the form of {bundleName}.bundle/{imageName} (mine was passing the bundlePath through every method
- Only bundles included in the app are searched (mine searches all bundles in all included libraries)

@spikebrehm We actually have an asset bundle in a CocoaPod which we are using; it does look like it's included in the main bundle though so this should work.

**Comment by mpretty-homepass** [comment]
@javache Did you want to open a separate PR or did you want me to update this PR with your solution?

**Comment by javache** [comment]
I'll land this internally and close this PR then. Thanks for reviewing that patch and building the original implementation!

**Comment by mpretty-homepass** [comment]
No worries.

**Comment by mkonicek** [comment]
Assigning to @javache to close this PR once it's landed internally - thanks for doing that!

**Comment by javache** [comment]
Yup, was done in https://github.com/facebook/react-native/commit/46b54fd7a80d6c8f422c7d2125733c18a20c7d7d
